### PR TITLE
SSE: Warn on dropped items in Union in Math Operation

### PR DESCRIPTION
**Pull request by kylebrandt**
**What is this feature?**

Unions in SSE operations currently drop items when doing a binary operation (e.g. $A + $B) from either A or B when there isn't a label "match" between them.

This is confusing, so to aid in this, we warn and display the count of items that where dropped, and which items they were.

image

image

Fixes #

**Special notes for your reviewer:**

Design wise, we really need to revisit "Union". In short, Math should probably have a dropdown that changes what happens with a binary operation (e.g. `$a OP $b`) in terms of union/join set operations based on labels. So this is more in the direction of "Informational based on the currently (and only mode)" so as not to create a future problem when tackling this in a broader sense. 

The output format is pretty dense. The way it works is from Left to Right, so - `((A union B) union C)`. So things can be dropped between union of A and B, as well as between the result of of the the union between A and B then having a Union with C. To put it another way, with `$A + $B + $C`, if there is a match between A and C, but not between A and B, the item will be dropped.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/expr/mathexp/parse/node.go [hunk 1]
```diff
@@ -77,6 +77,8 @@ func (t NodeType) String() string {
 		return "NodeString"
 	case NodeNumber:
 		return "NodeNumber"
+	case NodeVar:
```
Comment (kylebrandt) [inline pkg/expr/mathexp/parse/node.go:80]: random fix, can go in another PR if this isn't merged

File: pkg/expr/mathexp/exp.go [hunk 2]
```diff
@@ -198,25 +202,35 @@ type Union struct {
...
 	}
 	if aValueLen == 1 || bValueLen == 1 {
 		if aResults.Values[0].Type() == parse.TypeNoData || bResults.Values[0].Type() == parse.TypeNoData {
-			unions = append(unions, &Union{
+			appendUnions(&Union{
 				Labels: nil,
 				A:      aResults.Values[0],
 				B:      bResults.Values[0],
 			})
+			// TODO: Drop Count
```
Comment (itsmylife) [inline pkg/expr/mathexp/exp.go:226]: I assume this PR is still in progress, isn't it?

Comment (kylebrandt) [reply pkg/expr/mathexp/exp.go]: Fixed TODO, thanks forgot that one.

File: pkg/expr/mathexp/exp.go [hunk 3]
```diff
@@ -198,25 +202,63 @@ type Union struct {
...
+
+	aVar := biNode.Args[0].String()
+	bVar := biNode.Args[1].String()
+
+	aMatched := make([]bool, len(aResults.Values))
+	bMatched := make([]bool, len(bResults.Values))
+	collectDrops := func() {
+		check := func(v string, matchArray []bool, r *Results) {
+			for i, b := range matchArray {
+				if !b {
```
Comment (yuri-tceretian) [inline pkg/expr/mathexp/exp.go:219]: inversions are usually hard to read and missed. I think in this case you can invert it and interrupt the loop.

File: pkg/expr/mathexp/exp.go [hunk 4]
```diff
@@ -198,25 +202,63 @@ type Union struct {
...
 	unions := []*Union{}
+	appendUnions := func(u *Union) {
+		unions = append(unions, u)
+	}
+
+	aVar := biNode.Args[0].String()
+	bVar := biNode.Args[1].String()
+
+	aMatched := make([]bool, len(aResults.Values))
+	bMatched := make([]bool, len(bResults.Values))
```
[the remark this answers is not in the document]
Comment (kylebrandt) [reply pkg/expr/mathexp/exp.go:215]: This is what I tried at first but ran into difficulty. With the nested loop, we loop over B results multiple times. So during that loop - it is not known if a result will be matched in a future loop. So I ended up creating bool mask where I mark as matched during the loop, and then take the unmatched items.

**Review by yuri-tceretian — APPROVED** [review]
LGTM.
Not in this PR, but I think it would be great to have different modes of SSE evals: normal and diagnostics. In the diag mode each node could emit additional information that could be exposed to front-end and presented in some fancy UI. In normal mode it can be evaluated by alerting, so it does not have to spend time on collecting diag info.

**Comment by bobemoe** [comment]
I think these are erroneously appearing on my Graphite Math. Any way to hide them? See above linked issue for details.

**Comment by LutzFassl** [comment]
Please introduce option to hide those. I get those two annoying things now that I want to hide. They were not there in v9

in v9 I was happy with the behavior when running math expressions on a "no data" query it just failed silently which is ok for or should at least be a user setting. The "failed to fetch" and "attention symbol" is very annoying now!

image
